### PR TITLE
Feature/rafi/create an endpoint to enroll into a verified course

### PR DESCRIPTION
**Pull request by rafipriatna**
#### Step to reproduce:
HIT POST /courses/:courseId/enroll
Authentication: Access

#### See result:
IF course already enrolled:
Status Code: BAD REQUEST
```json
{
    "success": false,
    "message": "Course already enrolled."
}
```

IF course not verified:
Status Code: BAD REQUEST
```json
{
    "success": false,
    "message": "Course not verified."
}
```

IF success enroll course:
Status Code: OK
```json
{
    "success": true,
    "message": "Successfully enrolled a new course"
}
```

### CODE REVIEW

File: src/services/courseEnrollment.service.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                    'Course not verified.',
+                    StatusCodes.BAD_REQUEST);
+            }
+
+            const enrollCourse = CourseEnrollment.create(
+                { userId, courseId });
+            await CourseEnrollment.save(enrollCourse);
+        } catch (error) {
+            throw Errors.SERVER;
+        }
```
Comment (rahmatsyifana12) [inline src/services/courseEnrollment.service.ts]: ini gaperlu pake try catch dan throw error.SERVER di catchnya ya, udah dihandle di error.middleware

File: src/controllers/courseEnrollment.controller.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import { validate } from '../utils/validate.util';
+import {
+    courseEnrollmentSchema
+} from '../validations/courseEnrollment.validate';
+
+class CourseEnrollmentController {
+
+    async enrollNewCourse(req: Request, res: Response) {
+        const userPayload = await authService.getTokenPayload(req, 'ACCESS');
+        const courseId = validate(req, courseEnrollmentSchema, 'body');
```
Comment (rahmatsyifana12) [inline src/controllers/courseEnrollment.controller.ts]: ini courseId nya pake params aja, biar diroutingnya kita taro courseIdnya as params, misal:
/courses/:courseId/enrollment

File: src/routes/index.ts [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
 
 const router = Router();
 
+// Auth
 router.post('/auth/register', authController.register);
 router.post('/auth/login', authController.login);
 router.post('/auth/refresh', authenticate('ACCESS'), authController.refresh);
 
+// Course Enrollments
+router.post('/course/enroll', authenticate('ACCESS'),
```
Comment (rahmatsyifana12) [inline src/routes/index.ts]: tambahin courseId as params ya, taronya setelah /courses. coursenya ubah jadi courses, biar pake plural

**Review by rahmatsyifana12 — CHANGES_REQUESTED** [review]

File: src/routes/index.ts (hunk 3 above)
Comment (rafipriatna) [reply src/routes/index.ts]: Oke, berarti tetep POST request tanp body ya? Apa ganti ke GET?

Comment (rahmatsyifana12) [reply src/routes/index.ts]: tetep pake POST tanpa body

File: src/services/courseEnrollment.service.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                'Course already enrolled.',
+                StatusCodes.BAD_REQUEST);
+        }
+
+        const course = await Course.findOneBy({ id: courseId });
+        if (course?.isVerified) {
+            throw new ResponseError(
+                'Course not verified.',
+                StatusCodes.BAD_REQUEST);
+        }
```
Comment (rahmatsyifana12) [inline src/services/courseEnrollment.service.ts]: ini ngecek kalo is verified nya false ya? berarti kurang pake ! ga sih?

**Review by rahmatsyifana12 — CHANGES_REQUESTED** [review]

File: src/services/courseEnrollment.service.ts (hunk 4 above)
Comment (rafipriatna) [reply src/services/courseEnrollment.service.ts]: Itu pake `?` takutnya coursenya ga ada. Misalnya ada yg input courseId secara explisit ke endpointnya.

Apa di TypeORM udah otomatis ngehandle return response not found?

Oh iya, kurang `!` di depan wkwkk

**Review by rahmatsyifana12 — APPROVED** [review]

**Review by rahmatsyifana12 — APPROVED** [review]

**Pull request merged**